File: clients/python/model/player_message_game.py
class PlayerMessageGame:
    @staticmethod
    def read_from(stream):
        discriminant = stream.read_int()
        if discriminant == CustomDataMessage.TAG:
            return PlayerMessageGame.CustomDataMessage.read_from(stream)
        if discriminant == ActionMessage.TAG:
            return PlayerMessageGame.ActionMessage.read_from(stream)
        raise Exception("Unexpected discriminant value")

from .custom_data import CustomData
class CustomDataMessage(PlayerMessageGame):
    TAG = 0
    def __init__(self, data):
        self.data = data
    @staticmethod
    def read_from(stream):
        data = CustomData.read_from(stream)
        return CustomDataMessage(data)
    def write_to(self, stream):
        stream.write_int(self.TAG)
        self.data.write_to(stream)
    def __repr__(self):
        return "CustomDataMessage(" + \
            repr(self.data) + \
            ")"
PlayerMessageGame.CustomDataMessage = CustomDataMessage
from .versioned import Versioned
class ActionMessage(PlayerMessageGame):
    TAG = 1
    def __init__(self, action):
        self.action = action
    @staticmethod
    def read_from(stream):
        action = Versioned.read_from(stream)
        return ActionMessage(action)
    def write_to(self, stream):
        stream.write_int(self.TAG)
        self.action.write_to(stream)
    def __repr__(self):
        return "ActionMessage(" + \
            repr(self.action) + \
            ")"
PlayerMessageGame.ActionMessage = ActionMessage

File: clients/ruby/model/player_message_game.rb
require_relative 'custom_data'
require_relative 'versioned'
class PlayerMessageGame
    def self.read_from(stream)
        discriminant = stream.read_int()
        if discriminant == PlayerMessageGame::CustomDataMessage::TAG
            return PlayerMessageGame::CustomDataMessage.read_from(stream)
        end
        if discriminant == PlayerMessageGame::ActionMessage::TAG
            return PlayerMessageGame::ActionMessage.read_from(stream)
        end
        raise "Unexpected discriminant value"
    end

    class CustomDataMessage
        TAG = 0
        attr_accessor :data
        def initialize(data)
            @data = data
        end
        def self.read_from(stream)
            data = CustomData.read_from(stream)
            CustomDataMessage.new(data)
        end
        def write_to(stream)
            stream.write_int(TAG)
            @data.write_to(stream)
        end
    end
    class ActionMessage
        TAG = 1
        attr_accessor :action
        def initialize(action)
            @action = action
        end
        def self.read_from(stream)
            action = Versioned.read_from(stream)
            ActionMessage.new(action)
        end
        def write_to(stream)
            stream.write_int(TAG)
            @action.write_to(stream)
        end
    end
end
